### PR TITLE
Add a class for comparing versions of the record layer

### PR DESCRIPTION
**Pull request by ScottDugas**
This introduces a class for parsing and comparing versions of the Record Layer. 
This can be combined with the work in https://github.com/FoundationDB/fdb-record-layer/pull/3031 to control what versions a yaml test will run against.

### CODE REVIEW

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    private static final String ALPHANUMERIC_PATTERN = "\\d*[a-zA-Z-][0-9a-zA-Z-]";
+    private static final String PRE_RELEASE_PART = "(?:" + NUMBER_PATTERN + "|" + ALPHANUMERIC_PATTERN + ")";
+    private static final String BUILD_PART = "(?:\\d+|" + ALPHANUMERIC_PATTERN + ")";
+    private static final Pattern VERSION_PATTERN = Pattern.compile(
+            "^(?<versionNumbers>" + NUMBER_PATTERN + "(?:\\." + NUMBER_PATTERN + ")*)" +
+                    "(?:-(?<prerelease>" + dotSeparated(PRE_RELEASE_PART) + "))?" +
+                    "(?:\\+(?<buildMetadata>" + dotSeparated(BUILD_PART) + "))?$");
+    private static final Pattern NUMERIC_PATTERN = Pattern.compile("\\d+");
+
+    private final List<Integer> versionNumbers;
```
Comment (ohadzeliger) [inline yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java:61]: Some explanation of the `prerelease` and `metadata` parts (with example) could be useful.

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final SemanticVersion that = (SemanticVersion)o;
+        return Objects.equals(versionNumbers, that.versionNumbers) && Objects.equals(prerelease, that.prerelease) && Objects.equals(buildMetadata, that.buildMetadata);
```
Comment (ohadzeliger) [inline yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java:116]: This is theoretical, but is there meaning to the order of prerelease strings?

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            return versionComparison;
+        }
+        return comparePrerelease(o);
+    }
+
+    private int compareVersionNumbers(SemanticVersion o) {
+        // semver only allows 3, we only use 4.
+        // gradle supports an arbitrary number, but it gets complicated if there is a pre-release or build info
+        if (versionNumbers.size() != o.versionNumbers.size()) {
+            throw new IllegalArgumentException("Versions only differ by part count" + this + " vs " + o);
```
Comment (ohadzeliger) [inline yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java:138]: Have we considered treating "missing" version elements as `0`?

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        }
+        int length = Math.min(prerelease.size(), o.prerelease.size());
+        for (int i = 0; i < length; i++) {
+            String thisPart = prerelease.get(i);
+            String otherPart = o.prerelease.get(i);
+            if (NUMERIC_PATTERN.matcher(thisPart).matches()) {
+                if (NUMERIC_PATTERN.matcher(otherPart).matches()) {
+                    int thisNumber = Integer.parseInt(thisPart);
+                    int otherNumber = Integer.parseInt(otherPart);
+                    if (thisNumber < otherNumber) {
```
Comment (ohadzeliger) [inline yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java:165]: It may add extra cost of boxing and unboxing, but maybe `Objects.compare()` here may be clearer?

Comment (ohadzeliger) [reply yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java:165]:
```
int result = Objects.compare(thisNumber, otherNumber);
if (result != 0) {
    result result;
}
```

File: yaml-tests/src/test/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersionTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            "5.8.9, 5.8.9",
+            "9.7.4.3.0, 9.7.4.3.0",
+            "5.4.3-SNAPSHOT, 5.4.3-SNAPSHOT"
+    })
+    void equal(String rawVersionA, String rawVersionB) {
+        final SemanticVersion versionA = SemanticVersion.parse(rawVersionA);
+        final SemanticVersion versionB = SemanticVersion.parse(rawVersionB);
+        Assertions.assertAll(
+                () -> assertEquals(0, versionA.compareTo(versionB)),
+                () -> assertEquals(0, versionA.compareTo(versionB)));
```
Comment (ohadzeliger) [inline yaml-tests/src/test/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersionTest.java:47]:
```suggestion
                () -> assertEquals(0, versionB.compareTo(versionA)));
```

File: yaml-tests/src/test/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersionTest.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    @ParameterizedTest
+    @CsvSource({
+            "3, 4",
+            "4, 3",
+            "4, 4",
+    })
+    void alignsWithComparator(int versionA, int versionB) {
+        // Ensure that the signage aligns with the standard `compare` methods
+        // the value doesn't have to, but it does, and this is easier than asserting that they have the same sign
+        assertEquals(Integer.compare(versionA, versionB),
```
Comment (ohadzeliger) [inline yaml-tests/src/test/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersionTest.java:80]: Can't we use signum()?
```
assertEquals(
    Integer.compare(a, b).signum(), 
    SemanticVersion.parse(versionA + ".0.0").compareTo(
                         SemanticVersion.parse(versionB + ".0.0")).signum();
```

File: yaml-tests/src/test/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersionTest.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            "3.2-SNAPSHOT, 3.2.1",
+    })
+    void incomparible(String rawVersionA, String rawVersionB) {
+        final SemanticVersion versionA = SemanticVersion.parse(rawVersionA);
+        final SemanticVersion versionB = SemanticVersion.parse(rawVersionB);
+        Assertions.assertAll(
+                () -> assertThrows(IllegalArgumentException.class,
+                        () -> versionA.compareTo(versionB)),
+                () -> assertThrows(IllegalArgumentException.class,
+                        () -> versionA.compareTo(versionB)));
```
Comment (ohadzeliger) [inline yaml-tests/src/test/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersionTest.java:101]:
```suggestion
                        () -> versionB.compareTo(versionA)));
```

**Review by ohadzeliger — CHANGES_REQUESTED** [review]

**Comment by ohadzeliger** [comment]
Maybe add some tests with `parse(null)` and `compare to(null)`?

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java [hunk 8]
```diff
@@ -43,32 +44,47 @@
  *     Thus, this spec implements only slightly more than what the Record Layer needs, which is to say:
  *     a sequence of positive numbers ({@code 0|[1-9]\\d*} separated by {@code .}s, and an optional {@code -SNAPSHOT}
  *     at the end. Each version component is compared as integers, and versions with a different number of components
- *     are not comparable. A version with {@code -SNAPSHOT} version added to the less than the same set of numbers
- *     without {@code -SNAPSHOT}.
+ *     are not comparable. A version with {@code -SNAPSHOT} version added to a version the less than the same set of
+ *     numbers without {@code -SNAPSHOT} (e.g. {@code 3.4.5.1-SNAPSHOT < 3.4.5.1-SNAPSHOT}).
```
Comment (ohadzeliger) [inline yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java:48]:
```suggestion
 *     numbers without {@code -SNAPSHOT} (e.g. {@code 3.4.5.1-SNAPSHOT < 3.4.5.1).
```
?

**Review by ohadzeliger — CHANGES_REQUESTED** [review]
Just one comment change

**Review by ohadzeliger — DISMISSED** [review]

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+ *         <li><a href="https://maven.apache.org/ref/3.3.3/maven-artifact/apidocs/org/apache/maven/artifact/versioning/ComparableVersion.html">maven's version class</a>,
+ *         which is also outlined on <a href="https://cwiki.apache.org/confluence/display/MAVENOLD/Versioning">their wiki</a>.
+ *     </ul>
+ *     Gradle and Maven are the most important, as this is a java project, but they have complicated semantics, and I
+ *     had trouble reasoning about how they interact with non-ascii characters, and they appear to disagree.
+ *     Thus, this spec implements only slightly more than what the Record Layer needs, which is to say:
+ *     a sequence of positive numbers ({@code 0|[1-9]\\d*} separated by {@code .}s, and an optional {@code -SNAPSHOT}
+ *     at the end. Each version component is compared as integers, and versions with a different number of components
+ *     are not comparable. A version with {@code -SNAPSHOT} version added to a version the less than the same set of
+ *     numbers without {@code -SNAPSHOT} (e.g. {@code 3.4.5.1-SNAPSHOT < 3.4.5.1).
```
Comment (ScottDugas) [inline yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java:48]:
```suggestion
 *     numbers without {@code -SNAPSHOT} (e.g. {@code 3.4.5.1-SNAPSHOT < 3.4.5.1}).
```

**Review by ohadzeliger — APPROVED** [review]
